json_serializable/lib/src/type_helpers/map_helper.dart

```diff
@@ -90,30 +93,58 @@ class MapHelper extends TypeHelper<TypeHelperContextWithConfig> {
         context.config.anyMap ? 'as Map' : 'as Map<String, dynamic>';
 
     String keyUsage;
-    if (isEnumKey) {
+
+    if (isEnum(keyArg)) {
       keyUsage = context.deserialize(keyArg, _keyParam).toString();
     } else if (context.config.anyMap && !_isObjectOrDynamic(keyArg)) {
       keyUsage = '$_keyParam as String';
     } else {
       keyUsage = _keyParam;
     }
 
+    final toFromString = _forType(keyArg);
+    if (toFromString != null) {
+      keyUsage = toFromString.deserialize(keyArg, keyUsage, false, true);
+    }
+
     return '($expression $mapCast)$optionalQuestion.map('
         '($_keyParam, $closureArg) => MapEntry($keyUsage, $itemSubVal),)';
   }
 }
 
+final _intString = ToFromStringHelper('int.parse', 'toString()', 'int');
+
+final _instances = {
+  bigIntString,
+  dateTimeString,
+  _intString,
+  uriString,
+};
+
+Iterable<String> get _allowedTypeNames => const [
+      'Object',
+      'dynamic',
```

Comment:
What is the implication of having `dynamic` here?

Reply:
I can have a field of type `Map<dynamic, ...>` and things will "just work". The incoming JSON map (which, of course, has a `String` key) will be used directly.